### ISSUE TITLE
Feature: BEGIN immediate/exclusive per transaction

### ISSUE THREAD

**Issue by zombiezen** [mattn/go-sqlite3#400]
Right now, the `_txlock` connection parameter determines which `BEGIN` statement is used to start a transaction.  I'd like to be able to pick this at a per-transaction level.  I don't think there's a way of exposing this functionality to `database/sql`, but it would be good to add an option there eventually.  I'm imagining adding:

```go
func (c *SQLiteConn) BeginWithType(t TransactionType) (driver.Tx, error)

type TransactionType int

const (
  Deferred TransactionType = iota
  Immediate
  Exclusive
)
```

Relevant SO post: https://stackoverflow.com/questions/42657639/begin-immediate-in-sqlx-or-database-sql-golang

**Comment by mattn** [comment]
database/sql has IsolationLevel below.

```go
const (
	LevelDefault IsolationLevel = iota
	LevelReadUncommitted
	LevelReadCommitted
	LevelWriteCommitted
	LevelRepeatableRead
	LevelSnapshot
	LevelSerializable
	LevelLinearizable
)
```
It should map those values into the SQLite's transaction types.

**Comment by zombiezen** [comment]
I don't think it works in this case.  [SQLite's docs][1] assert:

> Except in the case of shared cache database connections with PRAGMA read_uncommitted turned on, all transactions in SQLite show "serializable" isolation.

[1]: https://www.sqlite.org/isolation.html

**Comment by mattn** [comment]
go-sqlite3 is driver for database/sql. So you can't see the BeginWithType from database/sql.

**Comment by zombiezen** [comment]
That's actually something I'm curious about: why does this package export the types used to implement the driver?  Can they be used directly if an application wants to have SQLite3-specific code?

**Comment by mattn** [comment]
I don't make sure but It's possible. However, interfaces are related on database/sql or database/sq/driver.
For example. SQLiteConn#Exec should be passed arguments []driver.Value. So programmer should consider what the value should be passed, what value will be returned.

**Comment by gaillard** [comment]
Well there is https://golang.org/pkg/database/sql/#TxOptions now. But as mentioned its not "isolation level" that is requested but either BEGIN IMMEDIATE/DEFFERED etc. All sqlite transactions are serializable level as far as I understand.
Best I can see to make this possible is for applications have a few different initialized sql drivers at the pkg level that have their own connection hooks and maybe have a field on *sqlite3.SQLiteConn to set?
I would say just change _txlock in the connection hook, but it has to already be given to the sql.Open.

**Comment by win-t** [comment]
Hi, it's been a year now,

I just wondering if 
```
db.Exec("BEGIN EXCLUSIVE TRANSACTION;")
// do something
db.Exec("COMMIT;")
```
works?, I don't know how to test it my self, that is why I asked it here

**Comment by zombiezen** [comment]
Nope. It's blocked on the upstream database/sql issue.

**Comment by otoolep** [comment]
>Nope. It's blocked on the upstream database/sql issue.

Huh? Why wouldn't it work? Those command just get sent straight to SQLite, and the transaction is started.

**Comment by zombiezen** [comment]
@win-t Your final suggestion looks like it would work, yes.  It's probably fine to close out this issue, then.

The previous suggestions don't either you don't have the guarantee that the statements are being sent to the same connection, or the transaction would have already started.

**Comment by GJRTimmer** [comment]
@zombiezen is this something we should add to either a wiki page or to the README is so could you send a PR for documentation update ? Wpuld be nice not to loose this kind of information in a issue which will be closed over time.

**Comment by niemeyer** [comment]
I've just run through this need as well. The most elegant solution would be to have another field in _sql.TxOptions_ named _Locking_ or something similar. The feature shouldn't abuse the _Isolation_ field as that's orthogonal to the locking mode and may be required together.

With that said, there's a simple workaround:
```
func beginImmediate(db *sql.DB) (*sql.Tx, error) {
        tx, err := db.Begin()
        if err == nil {
                _, err = tx.Exec("ROLLBACK; BEGIN IMMEDIATE")
        }
        return tx, err
}
```

**Comment by michaellenaghan** [comment]
If `TxOptions.ReadOnly` is `true`, `modernc` always uses deferred transactions. Otherwise it uses the value of `_txlock`. See [here](https://gitlab.com/cznic/sqlite/-/blob/master/sqlite.go?ref_type=heads#L734). 

`mattn` could use the same approach.

At the moment, `mattn` can't support both read and write transactions on the same connection. If you set `_txlock` to `immediate` *all* transactions become write transactions. If you leave `_txlock` at its default, `deferred`, you risk immediate `SQLITE_BUSY` errors when a read transaction is upgraded to a write transaction. See [here](https://berthub.eu/articles/posts/a-brief-post-on-sqlite3-database-locked-despite-timeout/).

In the meantime, one possible workaround is to have separate pools of read and write connections, each configured with a different `_txlock` value.

**Comment by michaellenaghan** [comment]
Btw, I developed a rather extensive SQLite benchmark for Go implementations; it's [here](https://github.com/michaellenaghan/go-sqlite-bench/tree/main?tab=readme-ov-file).
